haskell.packages.*.ghcWithPackages: Fix on darwin

9999years opened the pull request:
Previously, `ghcWithPackages` would fail to build on macOS because package `.conf` files now contain a literal `${pkgroot}` (which of course does not exist):

```
$ nix log /nix/store/ifi5mqp57iyb95c3xlzrflgcypq8xzy5-ghc-9.10.1-with-packages.drv | tail
...
/nix/store/i5gjwj9cnd74f7hvrq677vvyfm6alcpr-ghc-9.10.1/lib/ghc-9.10.1/lib/package.conf.d:
package.cache: Keeping existing link to /nix/store/dbhaj1v41x6zbmanczjik2bhp0l7fnj3-rio-0.1.22.0/lib/ghc-9.10.1/lib/package.conf.d/package.cache
package.cache.lock: Keeping existing link to /nix/store/dbhaj1v41x6zbmanczjik2bhp0l7fnj3-rio-0.1.22.0/lib/ghc-9.10.1/lib/package.conf.d/package.cache.lock
...
find: '${pkgroot}/../lib/aarch64-osx-ghc-9.10.1': No such file or directory
```

We can see the problematic string in the `.conf` files:


```
$ grep -Poz "dynamic-library-dirs:\s*\K .+\n" /nix/store/s5xrmb3583fd5qjrqsg73fblaq46gqpd-ghc-9.10.1-with-packages/lib/ghc-9.10.1/lib/package.conf.d/*
...
/nix/store/s5xrmb3583fd5qjrqsg73fblaq46gqpd-ghc-9.10.1-with-packages/lib/ghc-9.10.1/lib/package.conf.d/unix-2.8.5.1-ce1b.conf.copy: ${pkgroot}/../lib/aarch64-osx-ghc-9.10.1
...
```

Therefore, we use `ghc-pkg --simple-output list` to get a list of `.conf` files and `ghc-pkg --simple-output field "$pkg" dynamic-library-dirs` to get the value of the field while expanding `${pkgroot}` prefixes.

Closes #415436

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

9999years committed:
haskell.packages.*.ghcWithPackages: Fix on darwin

Previously, `ghcWithPackages` would fail to build on macOS because
package `.conf` files now contain a literal `${pkgroot}` (which of
course does not exist):

```
$ grep -Poz "dynamic-library-dirs:\s*\K .+\n" /nix/store/s5xrmb3583fd5qjrqsg73fblaq46gqpd-ghc-9.10.1-with-packages/lib/ghc-9.10.1/lib/package.conf.d/*
...
/nix/store/s5xrmb3583fd5qjrqsg73fblaq46gqpd-ghc-9.10.1-with-packages/lib/ghc-9.10.1/lib/package.conf.d/unix-2.8.5.1-ce1b.conf.copy: ${pkgroot}/../lib/aarch64-osx-ghc-9.10.1
...
```

Therefore, we use `ghc-pkg --simple-output list` to get a list of
`.conf` files and `ghc-pkg --simple-output field "$pkg"
dynamic-library-dirs` to get the value of the field while expanding
`${pkgroot}` prefixes.

sternenseemann commented:
> now contain

What makes you say that? I think there always used to be some of those, probably from GHC?

9999years commented:
> What makes you say that?

These derivations used to build. I'm not sure what changed but presumably this code wouldn't have ever worked if those placeholders were in the paths.

sternenseemann commented:
What changed is https://github.com/NixOS/nixpkgs/pull/411741, I assume that the placeholder just silently failed before.

9999years commented:
Yeah, you just beat me to it, bisect found that as well.

9999years committed:
tests.haskell.ghcWithPackages: init

`ghcWithPackages` broke in #411741, partially because we don't have
tests for it. This adds some tests to make sure it keeps working in the
future. For now these tests will fail until #415436 is fixed.

This is split off of #415434.

Adds a regression test for https://github.com/NixOS/nixpkgs/pull/224542

wolfgangwalther reviewed (dismissed):
This passed the newly introduced test-suite and all manual tests that I did.

I still encourage @KireinaHoro @TravisWhitaker @endgame @janvogt to test the issues you had as well, to make sure we got this right from all angles this time.

endgame commented:
The following command gave me a working shell where I could import `microlens` stuff and run some simple expressions:
```
nom shell --impure --expr '(builtins.getFlake "github:9999years/nixpkgs/fix-ghc-with-packages").legacyPackages.aarch64-darwin.haskellPackages.ghcWithPackages(p: [p.microlens])'
```

janvogt commented:
For me starting a shell with a ghcWithPackages on aarch64-darwin works:

```
> nix develop --override-input nixpkgs github:9999years/nixpkgs/fix-ghc-with-packages
warning: not writing modified lock file of flake 'path:/Users/jan/devel/repro/ghc-with-packages-aarch64-darwin-broken':
• Updated input 'nixpkgs':
    'github:NixOS/nixpkgs/3e3afe5174c561dee0df6f2c2b2236990146329f?narHash=sha256-frdhQvPbmDYaScPFiCnfdh3B/Vh81Uuoo0w5TkWmmjU%3D' (2025-06-07)
  → 'github:9999years/nixpkgs/030221d20f944ec7b96787869d04a4c423706e48?narHash=sha256-Ddrk8Mv0Me76thgvnOjrTYYwAfF%2BxwOoVSRTfzR0tLI%3D' (2025-06-10)
Welcome to the development shell!
bash-5.2$ 
bash-5.2$ ghc --interactive
GHCi, version 9.8.4: https://www.haskell.org/ghc/  :? for help
ghci> import Lens.Micro
ghci> (1,2) ^. _2
2
ghci> :q
Leaving GHCi.
bash-5.2$ 
```

sternenseemann commented on pkgs/development/haskell-modules/with-packages-wrapper.nix:
```diff
@@ -154,14 +154,21 @@ else
-          fi
+          dynamicLibraryDirs=("${bootLibDir}")
+
+          for pkg in $($out/bin/ghc-pkg list --simple-output); do
```
You could cross compile a `withPackages` environment in which case this would fail. In practice, this is not an issue since cross compiling to darwin is not possible.
